File: modules/kscan-live-vto-native/android/src/main/java/expo/modules/kscanlivevtonative/LiveVtoSessionState.kt
package expo.modules.kscanlivevtonative

/**
 * Part B: the native-internal session lifecycle state machine.
 *
 * WHY THIS EXISTS SEPARATELY FROM `types/vtoLive.ts`'s `LiveVtoSessionState`.
 * The JS contract's session states (`INITIALIZING`/`READY`/`TRACKING`/...) are
 * what a UI renders, and `services/vto/vtoLiveSession.ts`'s reducer already
 * owns that vocabulary exhaustively -- this file does not duplicate it. What
 * was missing on the native side was the layer BELOW that: a state machine
 * enforcing which native COMMANDS (`start`/`pause`/`resume`/`stop`/
 * `loadGarment`/`switchGarment`/capture/`dispose`) may be accepted from which
 * native lifecycle position, independent of whatever the camera/perception
 * pipeline underneath is doing frame-to-frame. Two different questions:
 * "is a pause valid right now" (this file) vs. "is the customer currently
 * tracked" (the JS reducer, fed by the events this state machine's
 * transitions cause `LiveVtoNativeSession` to emit).
 *
 * PURE ON PURPOSE. Zero Android/Expo imports (enforced by
 * RuntimeBoundaryTest.theGeometryAndReplayStackHasNoAndroidDependencies,
 * which does not need an exception entry for this file), so the entire
 * required test matrix -- every valid transition, every invalid one, every
 * race ordering -- runs on the JVM with no device, exactly like the geometry
 * conformance stack. `LiveVtoNativeSession` is the thin, Android-dependent
 * driver that calls into this table and does not re-implement it.
 */
enum class LiveVtoSessionState {
  CREATED,
  GARMENT_LOADING,
  READY,
  STARTING,
  RUNNING,
  PAUSED,
  STOPPING,
  STOPPED,
  CAPTURING,
  DISPOSED,
  ERROR,
}

/**
 * One command the module-level bridge may ask the state machine to accept.
 * Deliberately mirrors `LIVE_VTO_COMMANDS` in types/vtoLive.ts one-to-one for
 * the commands that have lifecycle meaning; `capturePersonFrame` and
 * `capturePreview` share one `CAPTURE` command here because they differ only
 * in which bitmap source `LiveVtoNativeSession` reads, never in what the
 * state machine allows.
 */
enum class LiveVtoSessionCommand {
  START,
  LOAD_GARMENT,
  SWITCH_GARMENT,
  PAUSE,
  RESUME,
  STOP,
  CAPTURE,
  DISPOSE,
}

/**
 * An async completion arriving back into the machine. Kept distinct from
 * `LiveVtoSessionCommand` because a completion is never rejected the way a
 * command can be -- by the time one of these arrives, the generation check in
 * `LiveVtoNativeSession` has already decided whether it is stale. This table
 * only says what state a GENUINE (non-stale) completion moves to.
 */
enum class LiveVtoSessionCompletion {
  RUNTIME_READY,
  RUNTIME_FAILED,
  GARMENT_LOADED,
  GARMENT_LOAD_FAILED,
  STOPPED,
  CAPTURE_FINISHED,
  FATAL,
}

data class LiveVtoSessionCommandResult(val accepted: Boolean, val next: LiveVtoSessionState)
data class LiveVtoSessionCompletionResult(val next: LiveVtoSessionState)

/**
 * The transition table itself. A pure function of (current state, input) ->
 * next state -- no fields, no side effects, so a test can enumerate every
 * (state, command) pair exhaustively without constructing anything.
 */
object LiveVtoSessionMachine {

  /**
   * Which state a command resumes INTO once its async work completes, for
   * commands that pass through an intermediate state
   * (`CAPTURING`/`GARMENT_LOADING` while switching). Recorded at the moment
   * the command is accepted, not re-derived later, so a capture that started
   * while RUNNING resumes to RUNNING even if something else were racing --
   * ambiguity a "resume to whatever makes sense now" rule would have.
   */
  fun resumeStateFor(before: LiveVtoSessionState): LiveVtoSessionState = when (before) {
    LiveVtoSessionState.RUNNING, LiveVtoSessionState.PAUSED -> before
    else -> LiveVtoSessionState.RUNNING
  }

  /**
   * Applies one command. Returns `accepted = false` (and the UNCHANGED
   * current state) for every invalid transition -- an invariant test asserts
   * the state genuinely does not move, not merely that a boolean came back
   * false.
   */
  fun apply(current: LiveVtoSessionState, command: LiveVtoSessionCommand): LiveVtoSessionCommandResult {
    val next = when (command) {
      LiveVtoSessionCommand.START -> when (current) {
        LiveVtoSessionState.CREATED, LiveVtoSessionState.STOPPED, LiveVtoSessionState.ERROR -> LiveVtoSessionState.STARTING
        else -> null
      }
      LiveVtoSessionCommand.LOAD_GARMENT -> when (current) {
        LiveVtoSessionState.CREATED, LiveVtoSessionState.STARTING,
        LiveVtoSessionState.READY, LiveVtoSessionState.STOPPED -> LiveVtoSessionState.GARMENT_LOADING
        else -> null
      }
      LiveVtoSessionCommand.SWITCH_GARMENT -> when (current) {
        LiveVtoSessionState.RUNNING, LiveVtoSessionState.PAUSED, LiveVtoSessionState.READY -> LiveVtoSessionState.GARMENT_LOADING
        else -> null
      }
      LiveVtoSessionCommand.PAUSE -> when (current) {
        LiveVtoSessionState.RUNNING -> LiveVtoSessionState.PAUSED
        else -> null
      }
      LiveVtoSessionCommand.RESUME -> when (current) {
        LiveVtoSessionState.PAUSED -> LiveVtoSessionState.RUNNING
        else -> null
      }
      LiveVtoSessionCommand.STOP -> when (current) {
        LiveVtoSessionState.CREATED, LiveVtoSessionState.STOPPED -> current // idempotent no-op
        LiveVtoSessionState.DISPOSED -> null // terminal: stop after dispose is refused, not a no-op
        else -> LiveVtoSessionState.STOPPING
      }
      LiveVtoSessionCommand.CAPTURE -> when (current) {
        LiveVtoSessionState.RUNNING, LiveVtoSessionState.PAUSED -> LiveVtoSessionState.CAPTURING
        else -> null
      }
      LiveVtoSessionCommand.DISPOSE -> LiveVtoSessionState.DISPOSED // universal, always idempotent
    }
    return if (next == null) {
      LiveVtoSessionCommandResult(accepted = false, next = current)
    } else {
      LiveVtoSessionCommandResult(accepted = true, next = next)
    }
  }

  /**
   * Applies a genuine (already generation-checked) completion. Every branch
   * is total: an unexpected completion for the current state (e.g. a stray
   * GARMENT_LOADED while CREATED, which should never happen if callers only
   * report completions for work the machine actually accepted) leaves the
   * state unchanged rather than guessing.
   */
  fun complete(
    current: LiveVtoSessionState,
    completion: LiveVtoSessionCompletion,
    resumeTo: LiveVtoSessionState = LiveVtoSessionState.RUNNING,
  ): LiveVtoSessionCompletionResult {
    val next = when (completion) {
      LiveVtoSessionCompletion.RUNTIME_READY ->
        if (current == LiveVtoSessionState.STARTING) LiveVtoSessionState.RUNNING else current
      LiveVtoSessionCompletion.RUNTIME_FAILED ->
        if (current == LiveVtoSessionState.STARTING) LiveVtoSessionState.ERROR else current
      LiveVtoSessionCompletion.GARMENT_LOADED ->
        if (current == LiveVtoSessionState.GARMENT_LOADING) resumeTo else current
      LiveVtoSessionCompletion.GARMENT_LOAD_FAILED ->
        if (current == LiveVtoSessionState.GARMENT_LOADING) LiveVtoSessionState.ERROR else current
      LiveVtoSessionCompletion.STOPPED ->
        if (current == LiveVtoSessionState.STOPPING) LiveVtoSessionState.STOPPED else current
      LiveVtoSessionCompletion.CAPTURE_FINISHED ->
        if (current == LiveVtoSessionState.CAPTURING) resumeTo else current
      LiveVtoSessionCompletion.FATAL -> LiveVtoSessionState.ERROR
    }
    return LiveVtoSessionCompletionResult(next = next)
  }

  /** Every state a fresh session may legitimately visit -- used by tests to
   *  assert the (state x command) matrix above is exhaustive, not sampled. */
  val ALL_STATES: List<LiveVtoSessionState> = LiveVtoSessionState.entries
  val ALL_COMMANDS: List<LiveVtoSessionCommand> = LiveVtoSessionCommand.entries
}

/**
 * Native mirror of `LiveVtoGarmentDescriptor` in types/vtoLive.ts. Re-declared
 * (not imported -- there is nothing to import across the JS/native boundary),
 * same reasoning as `LiveVtoGarment.kt`'s re-declaration of the `.ksgarment`
 * contract: the SAME fields, the SAME three supported template families,
 * checked here so a malformed descriptor is refused before any asset work
 * starts, not discovered partway through it.
 *
 * `assetKey`/`assetId`/`assetVersion` (added alongside the governed asset
 * resolver, services/vto/vtoLiveGarment.ts's `resolveLiveGarment`) are now
 * REQUIRED, not optional. Before this, `performGarmentLoad`
 * (LiveVtoTestRenderView.kt) resolved EVERY validated descriptor to the SAME
 * bundled fixture ("n1b-fixture", hardcoded) -- see
 * docs/vto-live-bridge-contract.md §13.5. `assetKey` is what closes that:
 * it is the ONE new field `performGarmentLoad` reads to pick which bundled
 * folder to load, and it is checked here against
 * [SUPPORTED_ASSET_KEYS] -- an ALLOWLIST of real bundled directory names,
 * never a caller-supplied filesystem path -- so a descriptor cannot address
 * an arbitrary asset path (mission "resolver cannot use file:// arbitrary
 * path traversal"). A descriptor missing a valid, allowlisted `assetKey` is
 * refused here exactly like a missing `productRef` always was: this is what
 * makes "no silent fixture fallback" a property of parsing, not of caller
 * discipline -- the diagnostic `active`/`replay`/`perception`/`camera` View
 * Props remain a SEPARATE, still-available explicit-fixture path (they call
 * `loadFixture("n1b-fixture")` directly, never through this descriptor).
 */
data class LiveVtoGarmentDescriptor(
  val productRef: String,
  val imageUrl: String,
  val canonicalCategory: String,
  val templateFamily: String,
  val assetKey: String,
  val assetId: String,
  val assetVersion: String,
) {
  companion object {
    val SUPPORTED_TEMPLATE_FAMILIES = setOf("t-shirt", "simple-top", "sweater")

    /** Allowlisted bundled-asset directory names. Mirrors
     *  services/vto/vtoLiveGarmentRegistry.ts's LIVE_VTO_ASSET_KEY_ALLOWLIST
     *  -- re-declared, not imported, same reasoning as the rest of this
     *  file's re-declared TS contract. Cross-checked mechanically against
     *  the TS side by __tests__/vtoLiveGarmentRegistryParity.test.js and,
     *  from this side, by LiveVtoGarmentDescriptorTest.kt. */
    val SUPPORTED_ASSET_KEYS = setOf("n1b-fixture", "n1c-asym-fixture")

    /** Parses and validates an Expo-bridged `Map<String, Any?>` command
     *  argument. Returns null for anything malformed -- never throws, so a
     *  caller can turn that into whichever failure event/exception its own
     *  contract requires. */
    fun fromBridgeMap(raw: Map<String, Any?>?): LiveVtoGarmentDescriptor? {
      val productRef = raw?.get("productRef") as? String
      val imageUrl = raw?.get("imageUrl") as? String
      val canonicalCategory = raw?.get("canonicalCategory") as? String
      val templateFamily = raw?.get("templateFamily") as? String
      val assetKey = raw?.get("assetKey") as? String
      val assetId = raw?.get("assetId") as? String
      val assetVersion = raw?.get("assetVersion") as? String
      if (productRef.isNullOrBlank() || imageUrl.isNullOrBlank() || canonicalCategory.isNullOrBlank()) return null
      if (templateFamily !in SUPPORTED_TEMPLATE_FAMILIES) return null
      if (assetKey !in SUPPORTED_ASSET_KEYS) return null
      if (assetId.isNullOrBlank() || assetVersion.isNullOrBlank()) return null
      return LiveVtoGarmentDescriptor(
        productRef, imageUrl, canonicalCategory, templateFamily!!, assetKey!!, assetId, assetVersion,
      )
    }
  }
}

File: modules/kscan-live-vto-native/ios/Core/LiveVtoSessionState.swift
/// Part B: the native-internal session lifecycle state machine.
///
/// Field-for-field port of Android's `LiveVtoSessionState.kt` -- see that
/// file's header for why this is a SEPARATE vocabulary from
/// `types/vtoLive.ts`'s `LiveVtoSessionState` (the JS-facing states a UI
/// renders) rather than a duplicate of it: this layer answers "is this
/// native COMMAND valid right now", the JS reducer answers "what does the
/// customer see", and this layer's transitions are what cause
/// `LiveVtoRenderView`'s session methods to emit the events that reducer
/// consumes.
///
/// PURE ON PURPOSE. Zero UIKit/ExpoModulesCore/AVFoundation imports (this
/// file lives in `ios/Core/`, the SAME zero-platform-import SwiftPM target
/// `LiveVtoRuntimeBoundaryTests.testCoreHasNoPlatformDependencies` already
/// enforces), so the entire required test matrix runs via `swift test` with
/// no device, exactly like the geometry conformance stack.
public enum LiveVtoSessionState: Equatable {
  case created
  case garmentLoading
  case ready
  case starting
  case running
  case paused
  case stopping
  case stopped
  case capturing
  case disposed
  case error

  public static let allCases: [LiveVtoSessionState] = [
    .created, .garmentLoading, .ready, .starting, .running, .paused, .stopping, .stopped, .capturing, .disposed, .error,
  ]
}

/// Mirrors Android's `LiveVtoSessionCommand` one-to-one.
public enum LiveVtoSessionCommand: Equatable {
  case start
  case loadGarment
  case switchGarment
  case pause
  case resume
  case stop
  case capture
  case dispose

  public static let allCases: [LiveVtoSessionCommand] = [
    .start, .loadGarment, .switchGarment, .pause, .resume, .stop, .capture, .dispose,
  ]
}

/// Mirrors Android's `LiveVtoSessionCompletion` one-to-one.
public enum LiveVtoSessionCompletion: Equatable {
  case runtimeReady
  case runtimeFailed
  case garmentLoaded
  case garmentLoadFailed
  case stopped
  case captureFinished
  case fatal
}

public struct LiveVtoSessionCommandResult: Equatable {
  public let accepted: Bool
  public let next: LiveVtoSessionState
}

public struct LiveVtoSessionCompletionResult: Equatable {
  public let next: LiveVtoSessionState
}

/// The transition table. A pure function of (current state, input) -> next
/// state -- no fields, no side effects.
public enum LiveVtoSessionMachine {

  /// Which state a command resumes INTO once its async work completes, for
  /// commands that pass through an intermediate state (`capturing`/
  /// `garmentLoading` while switching). See Android's `resumeStateFor` --
  /// recorded at acceptance time, not re-derived later.
  public static func resumeState(for before: LiveVtoSessionState) -> LiveVtoSessionState {
    switch before {
    case .running, .paused: return before
    default: return .running
    }
  }

  /// Applies one command. `accepted == false` always carries back the
  /// UNCHANGED current state.
  public static func apply(_ current: LiveVtoSessionState, _ command: LiveVtoSessionCommand) -> LiveVtoSessionCommandResult {
    let next: LiveVtoSessionState?
    switch command {
    case .start:
      switch current {
      case .created, .stopped, .error: next = .starting
      default: next = nil
      }
    case .loadGarment:
      switch current {
      case .created, .starting, .ready, .stopped: next = .garmentLoading
      default: next = nil
      }
    case .switchGarment:
      switch current {
      case .running, .paused, .ready: next = .garmentLoading
      default: next = nil
      }
    case .pause:
      next = (current == .running) ? .paused : nil
    case .resume:
      next = (current == .paused) ? .running : nil
    case .stop:
      switch current {
      case .created, .stopped: next = current // idempotent no-op
      case .disposed: next = nil // terminal: refused, not a no-op
      default: next = .stopping
      }
    case .capture:
      switch current {
      case .running, .paused: next = .capturing
      default: next = nil
      }
    case .dispose:
      next = .disposed // universal, always idempotent
    }
    guard let resolved = next else {
      return LiveVtoSessionCommandResult(accepted: false, next: current)
    }
    return LiveVtoSessionCommandResult(accepted: true, next: resolved)
  }

  /// Applies a genuine (already generation-checked) completion.
  public static func complete(
    _ current: LiveVtoSessionState,
    _ completion: LiveVtoSessionCompletion,
    resumeTo: LiveVtoSessionState = .running
  ) -> LiveVtoSessionCompletionResult {
    let next: LiveVtoSessionState
    switch completion {
    case .runtimeReady:
      next = (current == .starting) ? .running : current
    case .runtimeFailed:
      next = (current == .starting) ? .error : current
    case .garmentLoaded:
      next = (current == .garmentLoading) ? resumeTo : current
    case .garmentLoadFailed:
      next = (current == .garmentLoading) ? .error : current
    case .stopped:
      next = (current == .stopping) ? .stopped : current
    case .captureFinished:
      next = (current == .capturing) ? resumeTo : current
    case .fatal:
      next = .error
    }
    return LiveVtoSessionCompletionResult(next: next)
  }
}

/// Native mirror of `LiveVtoGarmentDescriptor` in types/vtoLive.ts.
/// Re-declared, not imported -- same reasoning as `LiveVtoGarment.swift`'s
/// re-declaration of the `.ksgarment` contract.
public struct LiveVtoGarmentDescriptor: Equatable {
  public static let supportedTemplateFamilies: Set<String> = ["t-shirt", "simple-top", "sweater"]

  public let productRef: String
  public let imageUrl: String
  public let canonicalCategory: String
  public let templateFamily: String

  public init(productRef: String, imageUrl: String, canonicalCategory: String, templateFamily: String) {
    self.productRef = productRef
    self.imageUrl = imageUrl
    self.canonicalCategory = canonicalCategory
    self.templateFamily = templateFamily
  }

  /// Parses and validates an Expo-bridged `[String: Any]` command argument.
  /// Returns nil for anything malformed -- never throws.
  public static func fromBridgeMap(_ raw: [String: Any]?) -> LiveVtoGarmentDescriptor? {
    guard
      let productRef = raw?["productRef"] as? String, !productRef.isEmpty,
      let imageUrl = raw?["imageUrl"] as? String, !imageUrl.isEmpty,
      let canonicalCategory = raw?["canonicalCategory"] as? String, !canonicalCategory.isEmpty,
      let templateFamily = raw?["templateFamily"] as? String,
      supportedTemplateFamilies.contains(templateFamily)
    else { return nil }
    return LiveVtoGarmentDescriptor(productRef: productRef, imageUrl: imageUrl, canonicalCategory: canonicalCategory, templateFamily: templateFamily)
  }
}
